### PR TITLE
etcd-tester: delete/range with limit, clean up

### PR DESCRIPTION
**Pull request by gyuho**

### CODE REVIEW

**Comment by gyuho** [comment]
@xiang90 @heyitsanthony PTAL. Been running 3, 5 node cluster over 150 rounds with no failure.

File: tools/functional-tester/etcd-tester/stresser.go [hunk 1]
```diff
@@ -111,9 +112,11 @@ func newStressDelete(kvc pb.KVClient, keySuffixRange int) stressFunc {
 
 func newStressDeletePrefix(kvc pb.KVClient, keySuffixRange int) stressFunc {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: These should probably be renamed to newStressRangeInterval / newStressDeleteInterval(kvc, keySuffixRange, limit) since they're not prefixes any more.

File: tools/functional-tester/etcd-tester/stresser.go [hunk 2]
```diff
@@ -95,6 +95,7 @@ func newStressRangePrefix(kvc pb.KVClient, keySuffixRange int) stressFunc {
 		_, err := kvc.Range(ctx, &pb.RangeRequest{
 			Key:      []byte("foo"),
 			RangeEnd: []byte(fmt.Sprintf("foo%d", rand.Intn(keySuffixRange))),
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: I don't think this needs to be randomized any more since there's a limit? It can just be fmt.Sprintf("foo%d", keySuffixRange) plus a random limit

**Comment by gyuho** [comment]
@heyitsanthony PTAL. Cleaned up more.

File: tools/functional-tester/etcd-tester/cluster.go [hunk 3]
```diff
@@ -36,10 +36,12 @@ const (
 type cluster struct {
 	v2Only bool // to be deprecated
 
-	datadir              string
+	agentEndpoints       []string
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/cluster.go]: I removed this in the last clean up. Is there a reason it needs to be in this struct?

File: tools/functional-tester/etcd-tester/cluster.go [hunk 4]
```diff
@@ -143,7 +131,8 @@ func (c *cluster) Reset() error {
 	for i, m := range c.Members {
 		eps[i] = m.Endpoint
 	}
-	return c.bootstrap(eps)
+	c.agentEndpoints = eps
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/cluster.go]: Does this need to be updated? I think this code was here because the `agentEndpoints` was taken out of the cluster structure.

File: tools/functional-tester/etcd-tester/cluster.go (hunk 3 above)
Comment (gyuho) [inline tools/functional-tester/etcd-tester/cluster.go]: I will changed it back as an argument. No strong opinion. Thanks.

File: tools/functional-tester/etcd-tester/stresser.go [hunk 5]
```diff
@@ -109,11 +110,14 @@ func newStressDelete(kvc pb.KVClient, keySuffixRange int) stressFunc {
 	}
 }
 
-func newStressDeletePrefix(kvc pb.KVClient, keySuffixRange int) stressFunc {
+func newStressDeleteInterval(kvc pb.KVClient, keySuffixRange, limit int) stressFunc {
 	return func(ctx context.Context) error {
+		// mid-value to range within [0, keySuffixRange)
+		num1 := keySuffixRange / 2
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: this should be randomized since the Puts are randomized

**Comment by gyuho** [comment]
@heyitsanthony PTAL. Thanks!

File: tools/functional-tester/etcd-tester/stresser.go [hunk 6]
```diff
@@ -71,49 +71,63 @@ func (st *stressTable) choose() stressFunc {
 	return st.entries[idx].f
 }
 
-func newStressPut(kvc pb.KVClient, keySuffixRange, keySize int) stressFunc {
+func (s *stresser) newStressPut(kvc pb.KVClient) stressFunc {
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: There's no need to tie these to a specific stresser. Once the stressFunc is created, the input values are sealed. Otherwise, it implies that the stresser will occasionally update `s.keySuffixRange` after creating the stressFunc, which isn't the case.

File: tools/functional-tester/etcd-tester/stresser.go [hunk 7]
```diff
@@ -71,49 +71,63 @@ func (st *stressTable) choose() stressFunc {
...
 	}
 }
 
-func newStressDeletePrefix(kvc pb.KVClient, keySuffixRange int) stressFunc {
+func (s *stresser) newStressDeleteInterval(kvc pb.KVClient) stressFunc {
 	return func(ctx context.Context) error {
+		resp, _ := kvc.Range(ctx, &pb.RangeRequest{
+			Key:       []byte("foo"),
+			RangeEnd:  []byte(fmt.Sprintf("foo%d", s.keySuffixRange)),
+			SortOrder: pb.RangeRequest_ASCEND,
```
Comment (heyitsanthony) [inline tools/functional-tester/etcd-tester/stresser.go]: the keys will be ascend sorted by key name by default, iirc. Adding a `SortOrder` forces the `Range` to get the full range from the backend before applying the limit

Comment (gyuho) [inline tools/functional-tester/etcd-tester/stresser.go]: Right. Will fix and test this patch more in cluster. Thanks.

**Comment by xiang90** [comment]
Can we get this merged anytime soon? 

/cc @gyuho @heyitsanthony

**Comment by gyuho** [comment]
@xiang90

Since we added range/delete stresser, 5-node cluster is returning ErrTooManyRequests and triggering set health key timeout (with intensive loads).

Basically, I am just trying to tune the stresser giving more back-off sleeps or longer timeouts...

**Comment by xiang90** [comment]
that does not matter. let's get this done. they are separate issues.

**Comment by gyuho** [comment]
@xiang90 @heyitsanthony Agree. I removed changes for QPS and just kept changes for range limit with clean up. I will address other 5-node issues elsewhere.

PTAL.

Thanks.

File: tools/functional-tester/etcd-tester/cluster.go [hunk 8]
```diff
@@ -36,10 +36,11 @@ const (
 type cluster struct {
 	v2Only bool // to be deprecated
 
-	datadir              string
+	dataDir              string
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/cluster.go]: revert this?

File: tools/functional-tester/etcd-tester/main.go [hunk 9]
```diff
@@ -28,22 +28,29 @@ var plog = capnslog.NewPackageLogger("github.com/coreos/etcd", "etcd-tester")
...
-	datadir := flag.String("data-dir", "agent.etcd", "etcd data directory location on agent machine.")
-	stressKeySize := flag.Int("stress-key-size", 100, "the size of each key written into etcd.")
-	stressKeySuffixRange := flag.Int("stress-key-count", 250000, "the count of key range written into etcd.")
+	dataDir := flag.String("data-dir", "agent.etcd", "etcd data directory location on agent machine.")
+	stressKeySize := flag.Uint("stress-key-size", 100, "the size of each key written into etcd.")
+	stressKeySuffixRange := flag.Uint("stress-key-count", 250000, "the count of key range written into etcd.")
+	stressKeyRangeLimit := flag.Uint("stress-range-limit", 50, "maximum number of keys to range or delete.")
 	limit := flag.Int("limit", -1, "the limit of rounds to run failure set (-1 to run without limits).")
-	stressQPS := flag.Int("stress-qps", 10000, "maximum number of stresser requests per second.")
+	stressQPS := flag.Uint("stress-qps", 20000, "maximum number of stresser requests per second.")
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/main.go]: revert this?

File: tools/functional-tester/etcd-tester/main.go [hunk 10]
```diff
@@ -28,22 +28,29 @@ var plog = capnslog.NewPackageLogger("github.com/coreos/etcd", "etcd-tester")
...
+		stressQPS:            int(*stressQPS),
+		stressKeySize:        int(*stressKeySize),
+		stressKeySuffixRange: int(*stressKeySuffixRange),
+		stressKeyRangeLimit:  int(*stressKeyRangeLimit),
+	}
+	if err := clus.bootstrap(strings.Split(*endpointStr, ",")); err != nil {
 		plog.Fatal(err)
 	}
-	defer c.Terminate()
+	defer clus.Terminate()
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/main.go]: revert this?

File: tools/functional-tester/etcd-tester/stresser.go [hunk 11]
```diff
@@ -171,11 +185,11 @@ func (s *stresser) Stress() error {
 	kvc := pb.NewKVClient(conn)
 
 	var stressEntries = []stressEntry{
-		{weight: 0.7, f: newStressPut(kvc, s.KeySuffixRange, s.KeySize)},
-		{weight: 0.07, f: newStressRange(kvc, s.KeySuffixRange)},
-		{weight: 0.07, f: newStressRangePrefix(kvc, s.KeySuffixRange)},
-		{weight: 0.07, f: newStressDelete(kvc, s.KeySuffixRange)},
-		{weight: 0.07, f: newStressDeletePrefix(kvc, s.KeySuffixRange)},
+		{weight: 0.7, f: newStressPut(kvc, s.keySuffixRange, s.keySize)},
+		{weight: 0.075, f: newStressRange(kvc, s.keySuffixRange)},
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/stresser.go]: revert the change from 0.07 to 0.075?

**Comment by gyuho** [comment]
@xiang90 Just reverted changed as you suggested. PTAL.

**Comment by heyitsanthony** [comment]
lgtm
